src/Passes/docs/src/DiscussionDocs/ThoughtsOnQubitManagement.md

```diff
@@ -0,0 +1,527 @@
+
+In some ways, `qubit_allocate` and `qubit_release` are a better choice than a referenced implementation as it allows direct compatibility existing languages such as C++. In appendix we demonstrate that we build the necessary resource management for a C++ runtime using the two functions `qubit_allocate` and `qubit_release`, meaning that it is directly possible to build a "Q++" version of C++ (if you disregard the other opaque types).
+
+Notably, `Array` and other opaque QIR types, does not seem to be directly supportable C or C++ as they add qualifiers `struct` in front of the opaque typename:
+
+```language
+%struct.QubitId = type opaque
+```
+
+# Allocation and destruction
+
+In this section, we will consider some frontend aspects around how qubits are currently allocated in a couple of different languages. As a starting point we will consider Q# and C++ as potential frontends. We will first discuss potential ways people will want to manage their qubits lifetime, then look at how qubits are allocated in C# today and finally discuss how qubits could be managed in C++.
+
+## Possible lifetime management of qubits
+
+We can think of a few different life-time management scenarios:
+
+1. Scoped lifetime management
+2. Lifetime limited to that of a child scope
+3. Lifetime limited to the lifetime of a containing object
+4. Fully reference counted.
+
+So far we have only considered 1 and 4. Using allocators and deallocators (in a languge that support move semantics) allows us to implement all four scenarios.
+
+Forcing reference counting will not allow us to do 2 and only to some degree 3. Further, we would only be able to map back to 1 if the count adds up at compile time. While 4 provides the most secure scenario there are many cases where 1-3 are desirable at least in ordinary resource management scenarios.
+
+## Qubit allocation in Q#
+
+Currently Q# implements 1 and it is appropriate to briefly consider how it currently uses `aloocate` and `release` to do so: Acquiring qubit access in Q# is done through a `use` statement which assigns the qubit to the current scope:
+
+```
+    operation Main(): Unit
+    {
+        use q1 = Qubit();
+
+        /// ... The underlying qubit ID is reserved through
+        /// the body of this function ...
+
+
+        /// ... and released right before the exit of the scope.
+    }
+```
+
+The use statement creates a allocate instruction and release instructions are inserted at the exit of the scope. However, one issue that is encountered in the Q# language is that it becomes hard for the frontend to provide garantuees that the qubit does not escape the scope. A good example of a hard case is
+
+```
+  operation Test() : Unit
+  {
```

Comment:
```suggestion
  operation Test() : Unit {
```

Reply:
This document is unfit to go into main. Deleted.